### PR TITLE
Add type constraints blog post

### PR DESCRIPTION
**Pull request by jguitana**

### CODE REVIEW

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+layout: post
+title: Type constraints for code reuse and readability
+author: jguitana
+---
+
+The case for statically typed languages is so that we can prove our programs are right at compile time. Although our languages are not powerful enough to achieve this all the time, we certainly not always use their full capabilities. Our intuition should guide us to reduce implementation space and improve code reuse, allowing us to focus on small parts of a big problem. We can only do this if we choose the right abstraction in the form of Types, because these act as constraints on the implementation space, and if built correctly bring with them reusable and tested code.
+
+Let's take for example the `Int` and the `Nat` types. Int represents a number from `-x to x`. Nat represents a number from `0 to x`.
+
+*The crux of the question is*: Whenever we expect a natural number in some closure, is it different to pass an Int argument then it is passing a Nat argument?
```
Comment (nbardiuk) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: `different ... then` should be `different ... than`
or maybe rephrase to `is there difference to pass an Int or a Nat argument`

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+---
+layout: post
+title: Type constraints for code reuse and readability
+author: jguitana
+---
+
+The case for statically typed languages is so that we can prove our programs are right at compile time. Although our languages are not powerful enough to achieve this all the time, we certainly not always use their full capabilities. Our intuition should guide us to reduce implementation space and improve code reuse, allowing us to focus on small parts of a big problem. We can only do this if we choose the right abstraction in the form of Types, because these act as constraints on the implementation space, and if built correctly bring with them reusable and tested code.
```
Comment (vascorsd) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: I'd recommend to end the sentence at "int the form of Types.". Then "These act..." and "and"s I think don't have commas right before.

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+*The crux of the question is*: Whenever we expect a natural number in some closure, is it different to pass an Int argument then it is passing a Nat argument?
+
+We are free to make a choice here, and many of us will not think about Nat.
+
+Or maybe just take Int for performance reasons? Well, not all abstractions are free of cost, but the same reason we give for not using lower level languages applies - we want to develop software faster.
+
+Maybe we will use negative numbers in the future? Well, first the program must run in the present. Refactoring constraints is ok, the compiler will help.
+
+And will our program be correct if we do use Int? It might, but we will have to test the case when Int goes negative. This is what is meant by reducing implementation space. Using Nat achieves this, but Int will not. So we end up by having to write extra tests. We can see that in a dynamic language with no types, everything is unconstrained to some primitive types, thus all implementations must be tested.
```
Comment (vascorsd) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: Also don't think sentences can start with "And". Here should be ok to just remove it.

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+def fromNel[A](nel: NonEmptyList[A]): A =
+    nel.tail.foldLeft(nel.head)(merge)
+```
+Take a moment to appreciate that we removed the need for Option from the `fromNel` function implementation. We delayed the need to implement such behavior. This is due to List having an extra check compared to NonEmptyList: `isEmpty`. Our NonEmptyList function is more robust as it requires no testing for this case.
+
+Actually, isEmpty is equivalent to `NonEmptyList.fromList` which accepts a `List` and returns an `Option[NonEmptyList[A]]`. So anytime we are doing multiple isEmpty checks we should probably realize our program should be working with a NonEmptyList.
+
+Also note that `.filter` on a NonEmptyList will transform it back to a List, because it may remove all elements. So we should also consider consolidating filtering in our programs.
+
+As yet another example, let's say we are parsing some data into a case class and that data has a List field which is never empty.
```
Comment (vascorsd) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: I think we can remove the "As yet" here also. The text flows nicely without it. Seems almost as if rushing using these time pronouns or wtv :)

**Review by vascorsd — CHANGES_REQUESTED** [review]
Seems nice, clean and with a good argument. I'm for it. Must still check some sentences to make sure they language-check.

**Review by nbardiuk — CHANGES_REQUESTED** [review]

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+Also note that we can represent the empty list case with `Option[NonEmptyList[A]]` where None is the empty list.
+
+Now, we can choose to implement this simple fold using a List or a NonEmptyList.
+
+```
+def merge[A](x: A, y: A): A = ???
+
+def fromList[A](ls: List[A]): Option[A] =
+    ls.foldLeft(Option.empty[A])((b, a) => b.map(merge(_, a)))
```
Comment (nbardiuk) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: this foldLeft always returns None, even when list is not empty

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 2 above)
Comment (sdsgomes) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Having "and"'s after commas isn't incorrect, but I'd remove it in this case because IMO `if built correctly` should be between commas (which would mean a lot of commas).
Actually, I would rephrase the last part of the sentence for clarity:
` because these act as constraints on the implementation space and bring with them reusable and tested code, if built correctly.`

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
+---
+layout: post
+title: Type constraints for code reuse and readability
+author: jguitana
+---
+
+The case for statically typed languages is so that we can prove our programs are right at compile time. Although our languages are not powerful enough to achieve this all the time, we certainly not always use their full capabilities. Our intuition should guide us to reduce implementation space and improve code reuse, allowing us to focus on small parts of a big problem. We can only do this if we choose the right abstraction in the form of Types, because these act as constraints on the implementation space, and if built correctly bring with them reusable and tested code.
+
+Let's take for example the `Int` and the `Nat` types. Int represents a number from `-x to x`. Nat represents a number from `0 to x`.
```
Comment (sdsgomes) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: I'd keep using \`Int\`  and  \`Nat\`.

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+*The crux of the question is*: Whenever we expect a natural number in some closure, is it different to pass an Int argument then it is passing a Nat argument?
+
+We are free to make a choice here, and many of us will not think about Nat.
+
+Or maybe just take Int for performance reasons? Well, not all abstractions are free of cost, but the same reason we give for not using lower level languages applies - we want to develop software faster.
+
+Maybe we will use negative numbers in the future? Well, first the program must run in the present. Refactoring constraints is ok, the compiler will help.
```
Comment (sdsgomes) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: Please capitalize the word "ok".

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+And will our program be correct if we do use Int? It might, but we will have to test the case when Int goes negative. This is what is meant by reducing implementation space. Using Nat achieves this, but Int will not. So we end up by having to write extra tests. We can see that in a dynamic language with no types, everything is unconstrained to some primitive types, thus all implementations must be tested.
+
+We've made a quick mental exercise on the cases above and we can see downsides and upsides as with most things. The bottom line is that if we are not solving a C10k like problem, we are probably better off building a more readable, smaller, well constrained program.
+
+Let's now take a look at `List` and `NonEmptyList` (from cats library).
```
Comment (sdsgomes) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: I would add a hyperlink to cats documention on "cats" word.

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+Actually, isEmpty is equivalent to `NonEmptyList.fromList` which accepts a `List` and returns an `Option[NonEmptyList[A]]`. So anytime we are doing multiple isEmpty checks we should probably realize our program should be working with a NonEmptyList.
+
+Also note that `.filter` on a NonEmptyList will transform it back to a List, because it may remove all elements. So we should also consider consolidating filtering in our programs.
+
+As yet another example, let's say we are parsing some data into a case class and that data has a List field which is never empty.
+- `case class Unconstrained(ints: List[Int])`
+- `case class Constrained(nats: NonEmptyList[Nat])`
+
+Our parsing logic for `Constrained` will be more involved (unless it was built already, which is a common thing with json parsing for example), but as soon as we create the parser for this data, we gain in that we do not require extra checks anywhere else in our code. If this class is used multiple times, we get multiple checks for free.
```
Comment (sdsgomes) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: I'd capitalize "json" :)

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+With a more constrained type we avoid having to make an extra conditional check in possibly many places in our code. This effectively means less code and tests. Also, if these types are used often enough, it will improve readability for everyone since we make less assumptions.
```
Comment (sdsgomes) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: Please add a comma after `constrained type` (aka `constrained type, we avoid having`).

I would rephrase the last sentence to:

> Also, if these types are used often enough, readability will improve for everyone since we make less assumptions.

**Review by sdsgomes — COMMENTED** [review]
Overall, it's pretty good 👍 
I want to read again after the comments are addressed before approving.

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 5 above)
Comment (nbardiuk) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: You can have Nil-save tail with `drop(1)` then use `headOption` so it looks similar to `fromNel` example
```scala
ls.drop(1).foldleft(ls.headOption)((b, a) => b.map(merge(_, a)))
```

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 2 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Settled for a mix if you agree: `We can only do this if we choose the right abstraction in the form of Types. These will act as constraints on the implementation space, bringing with them reusable and tested code.`

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 3 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Removed this "And" and an "Or" two paragraphs before this one for consistency.

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 4 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Changed to `As another` only. Thanks!

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 1 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Changed to the last phrase you mention, thanks!

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 5 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Yes that last one works great. Are you OK if I add this PS at the end of the post?

"PS Actually, the `fromList` function is incorrect as it always returns None. I guess it needed testing. The correct implementation is: `ls.foldLeft(Option.empty[A])((b, a) => b.map(merge(_, a)))`. Thanks Nazarii Bardiuk for pointing that out!"

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 6 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: I thought about this too. But they are used quite often and everything becomes highlighted. I think it is better to highlight only when introducing for the first time. Does this make sense?

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 7 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Done, thanks!

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 8 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Agreed

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 5 above)
Comment (nbardiuk) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: No, reader will have no context what this PS is about. But thanks for suggesting this

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 9 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Agree

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 10 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Agreed and agreed, thanks!

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 5 above)
Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: I'll fix the function above with your implementation 

PS The PS is wrong lol

**Review by nbardiuk — APPROVED** [review]

**Review by vascorsd — APPROVED** [review]

File: _posts/2018-11-13-type-constraints-correctness-reuse.md (hunk 6 above)
Comment (sdsgomes) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Yes, that's why I didn't complain about this any further in the post.

File: _posts/2018-11-13-type-constraints-correctness-reuse.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+*The crux of the question is*: Whenever we expect a natural number in some closure, is there a difference to pass an Int argument or a Nat argument?
+
+We are free to make a choice here, and many of us will not think about Nat.
+
+Maybe just take Int for performance reasons? Well, not all abstractions are free of cost, but the same reason we give for not using lower level languages applies - we want to develop software faster.
+
+Maybe we will use negative numbers in the future? Well, first the program must run in the present. Refactoring constraints is Ok, the compiler will help.
+
+Will our program be correct if we do use Int? It might, but we will have to test the case when Int goes negative. This is what is meant by reducing implementation space. Using Nat achieves this, but Int will not. So we end up by having to write extra tests. We can see that in a dynamic language with no types, everything is unconstrained to some primitive types, thus all implementations must be tested.
```
Comment (sdsgomes) [inline _posts/2018-11-13-type-constraints-correctness-reuse.md]: `So we end up by having to write extra tests.` -> should be `So we end up having to write extra tests.` Remove the "by" please.

Comment (jguitana) [reply _posts/2018-11-13-type-constraints-correctness-reuse.md]: Removed

**Review by sdsgomes — APPROVED** [review]

**Review by sdsgomes — APPROVED** [review]
